### PR TITLE
 add string && TestSprintf BenchmarkSprintf #86

### PR DESCRIPTION
**Pull request by Liberxue**
-  add TestBck_String TestNewHTTPObj
-  replace String NewHTTPObj

### CODE REVIEW

File: cmn/string.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package cmn
+
+import "strings"
+
+func Sprintf(str ...string) string {
```
Comment (VirrageS) [inline cmn/string.go:5]: Could you add comment on how this is different than `fmt.Sprintf`?

File: cmn/string_test.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+// goos: darwin
+// goarch: amd64
+// pkg: github.com/NVIDIA/aistore/cmn
+// cpu: Intel(R) Core(TM) i7-8559U CPU @ 2.70GHz
+// BenchmarkPlus-8   	       7	 147251687 ns/op	1537847900 B/op	   10050 allocs/op
+// PASS
+// ok  	github.com/NVIDIA/aistore/cmn	2.373s
+func BenchmarkPlus(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+		plusString(10000, randomString(30))
```
Comment (VirrageS) [inline cmn/string_test.go:73]: Same comment here about prepartions. And also this benchmark is hard to compare as you do plus `10000` times. Which you don't in `Sprintf`. Could you make those 2 benchmarks comparable?

I would suggest doing something like:
```
b.Run(func(b *B) {
	for n := 0; n < b.N; n++ {
 		Sprintf(preparedStrings)
 	}
}
b.Run(func(b *B) {
	for n := 0; n < b.N; n++ {
 		plusString(preparedStrings)
 	}
}
```
then you are using the same string and then you can conclude something with that.

File: cmn/tests/bucket_test.go [hunk 3]
```diff
@@ -154,3 +156,92 @@ func TestParseBckObjectURI(t *testing.T) {
...
+			}
+			if got := b.String(); got != tt.want {
+				t.Errorf("Bck.String() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestNewHTTPObj(t *testing.T) {
+	type args struct {
```
Comment (VirrageS) [inline cmn/tests/bucket_test.go:213]: No need to separate struct.

File: cmn/tests/bucket_test.go [hunk 4]
```diff
@@ -154,3 +156,92 @@ func TestParseBckObjectURI(t *testing.T) {
 		tassert.Errorf(t, err != nil, "expected error for input: %s", test.uri)
 	}
 }
+
+func TestBck_String(t *testing.T) {
+	type fields struct {
+		Name     string
```
Comment (VirrageS) [inline cmn/tests/bucket_test.go:162]: No need to make those fields exported (you can lowercase them).

**Review by VirrageS — CHANGES_REQUESTED** [review]

File: cmn/string.go (hunk 1 above)
Comment (Liberxue) [reply cmn/string.go:5]: I have added a [comparison benchmark](https://github.com/NVIDIA/aistore/pull/89/commits/2e1d5a9f3091c7bf71ffacf9d81088d342bb8b95#diff-0251cbcd27a7168f0001d0b5cc9a39c0801bde1f4044780892ce71c02fbe315eR46)
```
// go test -bench="String$" -benchmem  -count=3 .
// BenchmarkTheSprintString-8       2238352               482.8 ns/op          3072 B/op          2 allocs/op
// BenchmarkTheSprintString-8       2614776               472.4 ns/op          3072 B/op          2 allocs/op
// BenchmarkTheSprintString-8       2610601               458.1 ns/op          3072 B/op          2 allocs/op
// BenchmarkPlusString-8             790348              1488 ns/op            9984 B/op          5 allocs/op
// BenchmarkPlusString-8             818689              1373 ns/op            9984 B/op          5 allocs/op
// BenchmarkPlusString-8             811623              1434 ns/op            9984 B/op          5 allocs/op
// BenchmarkSprintfString-8          503763              2257 ns/op           11190 B/op         17 allocs/op
// BenchmarkSprintfString-8          496207              2282 ns/op           11190 B/op         17 allocs/op
// BenchmarkSprintfString-8          474106              2385 ns/op           11190 B/op         17 allocs/op
// PASS
// ok      github.com/NVIDIA/aistore/cmn   12.339s
```

File: cmn/string_test.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+// BenchmarkPlusString-8             790348              1488 ns/op            9984 B/op          5 allocs/op
+// BenchmarkPlusString-8             818689              1373 ns/op            9984 B/op          5 allocs/op
+// BenchmarkPlusString-8             811623              1434 ns/op            9984 B/op          5 allocs/op
+// BenchmarkSprintfString-8          503763              2257 ns/op           11190 B/op         17 allocs/op
+// BenchmarkSprintfString-8          496207              2282 ns/op           11190 B/op         17 allocs/op
+// BenchmarkSprintfString-8          474106              2385 ns/op           11190 B/op         17 allocs/op
+// PASS
+// ok      github.com/NVIDIA/aistore/cmn   12.339s
+func BenchmarkTheSprintString(b *testing.B) {
+	b.ResetTimer()
```
Comment (VirrageS) [inline cmn/string_test.go:61]: Now the cases are static so no need to `b.ResetTimer()`

File: cmn/string_test.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	}
+}
+func BenchmarkPlusString(b *testing.B) {
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		plusString(cases...)
+	}
+}
+
+func BenchmarkSprintfString(b *testing.B) {
```
Comment (VirrageS) [inline cmn/string_test.go:73]:
```suggestion
func BenchmarkFmtSprintfString(b *testing.B) {
```

File: cmn/string_test.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+// BenchmarkTheSprintString-8       2610601               458.1 ns/op          3072 B/op          2 allocs/op
+// BenchmarkPlusString-8             790348              1488 ns/op            9984 B/op          5 allocs/op
+// BenchmarkPlusString-8             818689              1373 ns/op            9984 B/op          5 allocs/op
+// BenchmarkPlusString-8             811623              1434 ns/op            9984 B/op          5 allocs/op
+// BenchmarkSprintfString-8          503763              2257 ns/op           11190 B/op         17 allocs/op
+// BenchmarkSprintfString-8          496207              2282 ns/op           11190 B/op         17 allocs/op
+// BenchmarkSprintfString-8          474106              2385 ns/op           11190 B/op         17 allocs/op
+// PASS
+// ok      github.com/NVIDIA/aistore/cmn   12.339s
+func BenchmarkTheSprintString(b *testing.B) {
```
Comment (VirrageS) [inline cmn/string_test.go:60]:
```suggestion
func BenchmarkSprintfString(b *testing.B) {
```

**Review by VirrageS — APPROVED** [review]
Besides small changes LGTM

**Comment by VirrageS** [comment]
Also could you squash your commits and add meaningful commit message? That would be awesome :)

File: cmn/tests/string_test.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+// go test -benchmem -run=^$ -bench="String$" github.com/NVIDIA/aistore/cmn/tests -count=3 .
+// BenchmarkSprintfString-8       2238352               482.8 ns/op          3072 B/op          2 allocs/op
+// BenchmarkSprintfString-8       2614776               472.4 ns/op          3072 B/op          2 allocs/op
+// BenchmarkSprintfString-8       2610601               458.1 ns/op          3072 B/op          2 allocs/op
+// BenchmarkPlusString-8             790348              1488 ns/op            9984 B/op          5 allocs/op
+// BenchmarkPlusString-8             818689              1373 ns/op            9984 B/op          5 allocs/op
+// BenchmarkPlusString-8             811623              1434 ns/op            9984 B/op          5 allocs/op
+// BenchmarkSprintfString-8          503763              2257 ns/op           11190 B/op         17 allocs/op
+// BenchmarkSprintfString-8          496207              2282 ns/op           11190 B/op         17 allocs/op
+// BenchmarkSprintfString-8          474106              2385 ns/op           11190 B/op         17 allocs/op
```
[the remark this answers is not in the document]
Comment (Liberxue) [reply cmn/tests/string_test.go]: benchmark test comments have been replaced, check [details](https://github.com/NVIDIA/aistore/pull/89/commits/c8b1ee285e5c15b09552c77221d9d34d8cb3e5ec#diff-588cd713c094d1cf33c94c851c55d8571398b8fd30f9fb4930bf7832d7d995bfR)

**Pull request closed without merging**